streams/src/main/java/org/apache/kafka/streams/state/internals/AbstractTransactionalStore.java

```diff
@@ -0,0 +1,442 @@
+        final MergeKeyValueIterator iterator = new MergeKeyValueIterator(
+            tmpStore().reverseRange(from, to),
+            mainStore().reverseRange(from, to),
+            true,
+            openIterators);
+        openIterators.add(iterator);
+        return iterator;
+    }
+
+    @Override
+    public KeyValueIterator<Bytes, byte[]> all() {
+        final MergeKeyValueIterator iterator = new MergeKeyValueIterator(
+            tmpStore().all(), mainStore().all(), openIterators);
+        openIterators.add(iterator);
+        return iterator;
+    }
+
+    @Override
+    public KeyValueIterator<Bytes, byte[]> reverseAll() {
+        final MergeKeyValueIterator iterator = new MergeKeyValueIterator(
+            tmpStore().reverseAll(), mainStore().reverseAll(), true, openIterators);
+        openIterators.add(iterator);
+        return iterator;
+    }
+
+    @Override
+    public long approximateNumEntries() {
+        try {
+            return Math.addExact(tmpStore().approximateNumEntries(), mainStore().approximateNumEntries());
+        } catch (final ArithmeticException e) {
+            return Long.MAX_VALUE;
+        }
+    }
+
+    @Override
+    public Position getPosition() {
+        return position;
+    }
+
+    private void doCommit() {
+        try (final KeyValueIterator<Bytes, byte[]> it = tmpStore().all()) {
+            while (it.hasNext()) {
+                final KeyValue<Bytes, byte[]> kv = it.next();
+                mainStore().put(kv.key, fromUncommittedValue(kv.value));
+            }
+        }
+
+        truncateTmpStore();
+    }
+
+    @Override
+    public <R> QueryResult<R> query(final Query<R> query, final PositionBound positionBound,
+        final QueryConfig config) {
+        return StoreQueryUtils.handleBasicQueries(
+            query,
+            positionBound,
+            config,
+            this,
+            position,
+            context
+        );
+    }
+
+    @Override
+    public <PS extends Serializer<P>, P> KeyValueIterator<Bytes, byte[]> prefixScan(final P prefix,
+                                                                                    final PS prefixKeySerializer) {
+        final MergeKeyValueIterator iterator = new MergeKeyValueIterator(
+            tmpStore().prefixScan(prefix, prefixKeySerializer),
+            mainStore().prefixScan(prefix, prefixKeySerializer),
+            openIterators);
+        openIterators.add(iterator);
+        return iterator;
+    }
+
+    private static KeyValue<Bytes, byte[]> fromUncommittedKV(final KeyValue<Bytes, byte[]> kv) {
+        if (kv.value[0] == DELETION) {
+            return null;
+        } else {
+            final byte[] value = new byte[kv.value.length - 1];
+            System.arraycopy(kv.value, 1, value, 0, value.length);
+            return new KeyValue<>(kv.key, value);
+        }
+    }
+
+    private static byte[] fromUncommittedValue(final byte[] value) {
```

Comment:
Hmm, this is tricky. We could introduce a _third_ store, which stores uncommitted tombstones. For most use-cases, this store would be so lightweight that it could be in-memory. We could even use a heuristic to automatically choose the most appropriate type of store:

- On startup, if an on-disk RocksDB tombstone store exists, use that.
- Otherwise, use an in-memory store.
- If the in-memory store exceeds a configurable threshold of keys, a RocksDB store is created to replace it, populated with the in-memory store keys.

When querying, you'd query all 3 stores, using the tombstone store to omit results, provided the tombstones are newer than any values found in the other stores.

I know it introduces _yet another store_, but I believe the majority of use-cases are so light on deletes that they'd comfortably fit in-memory, and therefore add very little overhead.